### PR TITLE
Adds Snakemake and Driver

### LINKED ISSUE #48: Script to produce plots and results
This issue can be closed when a script is produced that contains the workflow for producing results. Some other things should also be updated alongside:
- [ ] Add a functionality to include a label for the features used

### PR DESCRIPTION
**Pull request by samgdotson**
This PR closes #48 by adding a Snakemake file and a ``driver.py`` file.

Other changes:
* Adds ``.npy`` files to ``.gitignore``. 
* Adds saving capabilities to various functions.
* fixes datasets 

More scenarios will be added to the Snakefile in future PRs, for now I just needed to get a basic implementation done.

### CODE REVIEW

File: driver.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+    if not os.path.isdir(target_folder):
+        os.mkdir(target_folder)
+
+    var = get_variable_name(datafile_name)
+
+    plt.suptitle(f"{VARIABLES[var]} Prediction with ESN", fontsize=21)
+    plt.title(param_string(params))
+    plt.ylabel("Energy [kWh]", fontsize=16)
+    # plt.xlabel(f"Hours since {df.index[0]}", fontsize=16)
```
Comment (robfairh) [inline driver.py:279]: why is this commented out? consider removing it.

File: driver.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    print(X_in.shape, len(X_in.shape))
+
+
+# =============================================================================
+# ESN Optimization
+# =============================================================================
+    MAX_TRAINLEN = int(len(xdf) - params['future'])
+    print(f"Maximum training length is {MAX_TRAINLEN}")
+
+    # pred = esn_prediction(X_in.T, params)
```
Comment (robfairh) [inline driver.py:186]: why is this commented out? consider removing it.

File: driver.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        xdf = df
+
+    # Get the training data
+    power = np.array(xdf.kw).astype('float64')
+    power_norm = np.linalg.norm(power)
+    data_norms.append(power_norm)
+    X_in.append(power / power_norm)
+
+    if sun_elevation is not None:
+        # print("Adding sun elevation")
```
Comment (robfairh) [inline driver.py:160]: why is this commented out? consider removing it.

File: driver.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+        if opt in ('-S', '--save_prefix'):
+            save_prefix = arg
+
+        if opt in ('-H'):
+            params['window'] = int(arg)
+
+    # Align the two dataframes
+    if wdf is not None:
+        # print('dataframes must be aligned')
```
Comment (robfairh) [inline driver.py:146]: why is this commented out? consider removing it.

File: driver.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+    try:
+        opts, args = getopt.getopt(sys.argv[1:],
+                                   'uwdpheH:i:f:oS:',
+                                   ['infile=', 'altfile', 'outfile=',
+                                    'save_prefix='])
+    except getopt.GetoptError:
+        print(f'Valid options are: {options_dict}')
+        sys.exit(1)
+    # print(args)
```
Comment (robfairh) [inline driver.py:105]: why is this commented out? consider removing it.

File: driver.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+plt.rcParams['figure.figsize'] = (12, 9)
+plt.rcParams['figure.edgecolor'] = 'k'
+plt.rcParams['figure.facecolor'] = 'w'
+plt.rcParams['savefig.dpi'] = 400
+plt.rcParams['savefig.bbox'] = 'tight'
+plt.rcParams['text.usetex'] = True
+plt.rcParams['font.family'] = "serif"
+
+# Optimization Sets
+# radius_set = [0.5, 0.7, 0.9, 1, 1.1,1.3,1.5]
```
Comment (robfairh) [inline driver.py:24]: why is this commented out? consider removing it.
or maybe add a comment on why that list may be preferred over the one on line 27

**Comment by robfairh** [comment]
Just a few minor comments. Also, there is a conflicting file.
Other than that, it looks good.

File: driver.py (hunk 6 above)
Comment (samgdotson) [reply driver.py]: I included shorter lists to execute a snakemake run in a reasonable amount of time.

File: loss_plots.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    yvar : string
+        The name of the second variable
+    figure_path : string
+    """
+    splitstring = fname.split('_')
+
+    xset, yset = None, None
+    xvar, yvar = None, None
+
+    # print(fname)
```
Comment (robfairh) [inline loss_plots.py:89]: It feels like you used this for debugging the script.
If the script has the desirable behavior, you should remove the line.

Comment (robfairh) [reply loss_plots.py:89]: The same comment applies to other commented prints.

File: tools.py [hunk 8]
```diff
@@ -19,11 +19,33 @@ def MSE(yhat, y):
     mse : float
         The mean squared error between yhat and y.
     '''
-    mse = np.sqrt(np.mean((yhat.flatten() - y.flatten())**2))
+    mse = np.sqrt(np.mean((y.flatten() - yhat.flatten())**2))
```
Comment (robfairh) [inline tools.py:22]: is y.flatten() always greater than yhat.flatten()?
You could add a np.absolute() if not.

Comment (samgdotson) [reply tools.py:22]: It doesn't matter for this one because it is squared

Comment (robfairh) [reply tools.py:22]: oh yes, you are right

**Comment by robfairh** [comment]
Merging

**Pull request merged**